marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo/Types.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Marconi.ChainIndex.Api.JsonRpc.Endpoint.Utxo.Types (
+  AddressUtxoResult (AddressUtxoResult),
+  GetUtxosFromAddressParams (GetUtxosFromAddressParams),
+  GetUtxosFromAddressResult (GetUtxosFromAddressResult),
+) where
+
+import Cardano.Api (FromJSON, ToJSON)
+import Cardano.Api qualified as C
+import Control.Applicative ((<|>))
+import Data.Aeson (FromJSON (parseJSON), (.:), (.:?))
+import Data.Aeson qualified as Aeson
+import GHC.Generics (Generic)
+import Marconi.ChainIndex.Api.JsonRpc.Endpoint.Utxo.SpentInfoResult (SpentInfoResult)
+import Marconi.ChainIndex.Api.JsonRpc.Endpoint.Utxo.Wrappers (UtxoTxInput, ValueWrapper)
+import Marconi.ChainIndex.Api.Types ()
+import Marconi.ChainIndex.Types (TxIndexInBlock)
+
+data GetUtxosFromAddressParams = GetUtxosFromAddressParams
+  { address :: !String
+  -- ^ Address to query for
+  , createdAtOrAfterSlotNo :: !(Maybe C.SlotNo)
+  -- ^ Upper slot in the window
+  , unspentBeforeSlotNo :: !(Maybe C.SlotNo)
+  -- ^ lower slot in the window
+  }
+  deriving (Show, Eq, Generic)
+
+instance FromJSON GetUtxosFromAddressParams where
+  parseJSON =
+    let
+      parse v = do
+        (lo, hi) <- buildInterval v
+        GetUtxosFromAddressParams
+          <$> v
+            .: "address"
+          <*> pure lo
+          <*> pure hi
+     in
+      Aeson.withObject "GetUtxosFromAddressParams" parse
+    where
+      buildInterval v = do
```

Comment:
I thought about this some more. I think this is more correct than a QueryError in the Queryable instance. The reason is, these constraints are detailed by the specification. As a result, I think this genuinely is a parsing error. Happy to talk more about this if you disagree.

Reply:
The constraints are detailed by the specification but they aren't enforced by the types themselves (in the specification).
Although, as we can enforce it on the Haskell side, it's probably the way to go.